TidepoolServiceKit/TidepoolService.swift

```diff
@@ -301,82 +310,49 @@ extension TidepoolService: RemoteDataService {
     }
 
     public var carbDataLimit: Int? { return 1000 }
 
-    public func uploadCarbData(created: [SyncCarbObject], updated: [SyncCarbObject], deleted: [SyncCarbObject], completion: @escaping (Result<Bool, Error>) -> Void) {
+    public func uploadCarbData(created: [SyncCarbObject], updated: [SyncCarbObject], deleted: [SyncCarbObject]) async throws {
         guard let userId = userId, let hostIdentifier = hostIdentifier, let hostVersion = hostVersion else {
-            completion(.failure(TidepoolServiceError.configuration))
-            return
+            throw TidepoolServiceError.configuration
         }
 
-        Task {
-            do {
-                let createdUploaded = try await createData(created.compactMap { $0.datum(for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion) })
-                let updatedUploaded = try await updateData(updated.compactMap { $0.datum(for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion) })
-                let deletedUploaded = try await deleteData(withSelectors: deleted.compactMap { $0.selector })
-                completion(.success(createdUploaded || updatedUploaded || deletedUploaded))
-            } catch {
-                completion(.failure(error))
-            }
-        }
+        let _ = try await createData(created.compactMap { $0.datum(for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion) })
+        let _ = try await updateData(updated.compactMap { $0.datum(for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion) })
+        let _ = try await deleteData(withSelectors: deleted.compactMap { $0.selector })
     }
 
     public var doseDataLimit: Int? { return 1000 }
 
-    public func uploadDoseData(created: [DoseEntry], deleted: [DoseEntry], completion: @escaping (_ result: Result<Bool, Error>) -> Void) {
+    public func uploadDoseData(created: [DoseEntry], deleted: [DoseEntry]) async throws {
         guard let userId = userId, let hostIdentifier = hostIdentifier, let hostVersion = hostVersion else {
-            completion(.failure(TidepoolServiceError.configuration))
-            return
+            throw TidepoolServiceError.configuration
         }
 
-        Task {
-            do {
-                let createdUploaded = try await createData(created.flatMap { $0.data(for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion) })
-                let deletedUploaded = try await deleteData(withSelectors: deleted.flatMap { $0.selectors })
-                completion(.success(createdUploaded || deletedUploaded))
-            } catch {
-                completion(.failure(error))
-            }
-        }
+        let _ = try await createData(created.flatMap { $0.data(for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion) })
+        let _ = try await deleteData(withSelectors: deleted.flatMap { $0.selectors })
```

Comment:
Similar to above. Another possible use of `TaskGroup`?

```suggestion
        await withThrowingTaskGroup(of: Void.self) { group in
            group.addTask { try await createData(created.flatMap { $0.data(for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion) }) }
            group.addTask { try await deleteData(withSelectors: deleted.flatMap { $0.selectors }) }
            await group.waitForAll()
        }
```

Reply:
Serialization is needed here; some of the deletes may be for objects created in the createData line.